src/raftstore/store/worker/read.rs

```diff
@@ -0,0 +1,895 @@
+            true
+        } else {
+            debug!(
+                "{} deny, applied_index_term {} != term {} ",
+                self.delegate.tag, self.delegate.applied_index_term, self.delegate.term
+            );
+            self.metrics.rejected_by_appiled_term += 1;
+            false
+        }
+    }
+
+    fn inspect_lease(&mut self) -> LeaseState {
+        // TODO: disable localreader if we did not enable raft's check_quorum.
+        if let Some(ref lease) = self.delegate.leader_lease {
+            // None means now.
+            if LeaseState::Valid == lease.inspect(None) {
+                return LeaseState::Valid;
+            }
+            self.metrics.rejected_by_lease_expire += 1;
+            debug!(
+                "{} leader lease is expired: {:?}",
+                self.delegate.tag, self.delegate.leader_lease
+            );
+        } else {
+            debug!("{} leader lease is None", self.delegate.tag,);
+            self.metrics.rejected_by_no_lease += 1;
+        }
+        LeaseState::Expired
+    }
+}
+
+impl<C: Sender<StoreMsg>> Runnable<Task> for LocalReader<C> {
+    fn run(&mut self, task: Task) {
+        fail_point!("local_reader_on_run");
+        match task {
+            Task::Register(delegate) => {
+                debug!(
+                    "{} register ReadDelegate for {:?}",
+                    delegate.tag, delegate.peer
+                );
+                self.delegates.insert(delegate.region.get_id(), delegate);
+            }
+            Task::Read(StoreMsg::RaftCmd {
+                send_time,
+                request,
+                callback,
+            }) => self.propose_raft_command(request, callback, send_time),
+            Task::Read(StoreMsg::BatchRaftSnapCmds {
+                batch, on_finished, ..
+            }) => self.propose_batch_raft_snapshot_command(batch, on_finished),
+            Task::Read(other) => {
+                unimplemented!("unsupported Msg {:?}", other);
+            }
+            Task::Update((region_id, progress)) => {
+                if let Some(delegate) = self.delegates.get_mut(&region_id) {
+                    delegate.update(progress);
+                } else {
+                    debug!(
+                        "unregistered ReadDelegate, region_id: {}, {:?}",
+                        region_id, progress
+                    );
+                }
+            }
+            Task::Destroy(region_id) => {
+                self.delegates.remove(&region_id);
+            }
+        }
+    }
+
+    fn on_tick(&mut self) {
+        self.metrics.flush();
+        let count = self.delegates.values().fold(0i64, |mut acc, delegate| {
+            if let Some(ref lease) = delegate.leader_lease {
+                if lease.term() == delegate.term {
+                    acc += 1;
+                }
+            }
+            acc
+        });
+        LOCAL_READ_LEADER.set(count);
+    }
+}
+
+struct ReadMetrics {
```

Comment:
Can this be optimized by static metrics?

Reply:
Actually it's faster than static metrics.